llvm/include/llvm/Transforms/Utils/LoopUtils.h

```diff
@@ -319,7 +319,7 @@ void addStringMetadataToLoop(Loop *TheLoop, const char *MDString,
 /// In addition if \p EstimatedLoopInvocationWeight is not null it is
 /// initialized with weight of loop's latch leading to the exit.
 /// Returns a valid positive trip count, or std::nullopt when a meaningful
-/// estimate cannot be made.
+/// estimate cannot be made (including when the trip count would overflow).
```

Comment:
More accurately stating that it won't fit in the result type?

```suggestion
/// estimate cannot be made (including when the trip count would not fit in the result type).
```

In general, the way this is handled could be improved I think, as we underestimate the trip counts if they get too large. It might make sense to return an estimate of the lower bound of the trip count here as a follow-up, assuming that the absolute value doesn't matter too much for values >= 2^32?

Reply:
> It might make sense to return an estimate of the lower bound of the trip count here as a follow-up, assuming that the absolute value doesn't matter too much for values >= 2^32?

I was thinking along the same lines. Might as well return UINT_MAX, when it would overflow, no?

Reply:
Yes that should be fine and improve results in general I think.